internal/era/era.go

```diff
@@ -0,0 +1,553 @@
+		}
+		return nil
+	}
+
+	// Write block data.
+	if err := snappyWrite(TypeCompressedHeader, header); err != nil {
+		return err
+	}
+	if err := snappyWrite(TypeCompressedBody, body); err != nil {
+		return err
+	}
+	if err := snappyWrite(TypeCompressedReceipts, receipts); err != nil {
+		return err
+	}
+	// Also write total difficulty, but don't snappy encode.
+	btd := bigToBytes32(td)
+	n, err := b.w.Write(TypeTotalDifficulty, btd[:])
+	b.written += n
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Finalize computes the accumulator and block index values, then writes the
+// corresponding e2store entries.
+func (b *Builder) Finalize() (common.Hash, error) {
+	if b.startNum == nil {
+		return common.Hash{}, fmt.Errorf("finalize called on empty builder")
+	}
+	// Compute accumulator root and write entry.
+	root, err := ComputeAccumulator(b.hashes, b.tds)
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("error calculating accumulator root: %w", err)
+	}
+	n, err := b.w.Write(TypeAccumulator, root[:])
+	b.written += n
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("error writing accumulator: %w", err)
+	}
+	// Get beginning of index entry to calculate block relative offset.
+	base := int64(b.written + (3 * 8)) // skip e2store header (type, length) and start block
+
+	// Construct block index. Detailed format described in Builder
+	// documentation, but it is essentially encoded as:
+	// "start | index | index | ... | count"
+	var (
+		count = len(b.indexes)
+		index = make([]byte, 16+count*8)
+	)
+	binary.LittleEndian.PutUint64(index, *b.startNum)
+	// Each offset is relative from the position it is encoded in the
+	// index. This means that even if the same block was to be included in
+	// the index twice (this would be invalid anyways), the relative offset
+	// would be different. The idea with this is that after reading a
+	// relative offset, the corresponding block can be quickly read by
+	// performing a seek relative to the current position.
+	for i, offset := range b.indexes {
+		relative := int64(offset) - (base + int64(i)*8)
+		binary.LittleEndian.PutUint64(index[8+i*8:], uint64(relative))
+	}
+	binary.LittleEndian.PutUint64(index[8+count*8:], uint64(count))
+
+	// Finally, write the block index entry.
+	if _, err := b.w.Write(TypeBlockIndex, index); err != nil {
+		return common.Hash{}, fmt.Errorf("unable to write block index: %w", err)
+	}
+
+	return root, nil
+}
+
+// writeVersion writes a version entry to e2store.
+func writeVersion(w *e2store.Writer) error {
+	_, err := w.Write(TypeVersion, nil)
+	return err
+}
+
+// Reader reads an Era1 archive.
+// See Builder documentation for a detailed explanation of the Era1 format.
+type Reader struct {
+	r        io.ReadSeeker
+	offset   uint64
+	metadata *metadata
+}
+
+// NewReader returns a new Reader instance.
+func NewReader(r io.ReadSeeker) *Reader {
+	// qq: does it make more sense to read the metadata here and
```

Comment:
Hmm usually I would leave it this way, but since you need to have two additional checks because of it, I kinda lean to removing it

Reply:
removed!